src/frontend/src/binder/expr/value.rs

```diff
@@ -113,12 +113,37 @@ impl Binder {
 
         Ok(literal)
     }
+
+    /// `ARRAY[...]` is represented as an function call at the binder stage.
+    pub(super) fn bind_array(
+        &mut self,
+        exprs: Vec<risingwave_sqlparser::ast::Expr>,
+    ) -> Result<ExprImpl> {
+        let mut exprs = exprs
+            .into_iter()
+            .map(|e| self.bind_expr(e))
+            .collect::<Result<Vec<ExprImpl>>>()?;
+        let element_type = align_types(exprs.iter_mut())?;
```

Comment:
When `exprs` is empty, the type will fallback to `varchar`.

In most cases, PG exclude `unknown` types when aligning types, and then uses `text` if all of them are `unknown`.

But if the whole list is empty, it reports an error:
```
test=# select array[];
ERROR:  cannot determine type of empty array
LINE 1: select array[];
               ^
HINT:  Explicitly cast to the desired type, for example ARRAY[]::integer[].

test=# select pg_typeof(array[null]);
 pg_typeof 
-----------
 text[]
(1 row)
```

Reply:
Good to know! I've tested and added these two cases in e2e. Our results are:
- If `array[]` is given, the query will fail at the parser stage.
- For `array[null]`, the query runs as expected, the same as postgres.

Reply:
Seems we have no way to create an empty array then?

Reply:
Support of `array[]::int[]` is being worked on here: #5402 (from #4715)